File: app/game/core/input/Controller.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf
local __TS__ArrayIncludes = ____lualib.__TS__ArrayIncludes
local ____exports = {}
local ____CONSTS = require("ecs.CONSTS")
local InputKeys = ____CONSTS.InputKeys
function ____exports.newController()
    local pressed, listeners, hashesKey
    function pressed(key, isPressed)
        if __TS__ArrayIncludes(hashesKey, key) then
            if listeners[__TS__ArrayIndexOf(hashesKey, key)] ~= nil then
                for ____, listener in ipairs(listeners[__TS__ArrayIndexOf(hashesKey, key)]) do
                    listener(isPressed)
                end
            end
        end
    end
    listeners = {}
    hashesKey = __TS__ArrayMap(
        InputKeys,
        function(____, k) return hash(k) end
    )
    local function input(action_id, action)
        if action_id ~= hash("touch") and (action.pressed or action.released) then
            pressed(action_id, action.pressed ~= nil and action.pressed == true)
        end
    end
    local function on(key, func)
        if listeners[__TS__ArrayIndexOf(InputKeys, key)] == nil then
            listeners[__TS__ArrayIndexOf(InputKeys, key)] = {}
        end
        local ____listeners___TS__ArrayIndexOf_result_0 = listeners[__TS__ArrayIndexOf(InputKeys, key)]
        ____listeners___TS__ArrayIndexOf_result_0[#____listeners___TS__ArrayIndexOf_result_0 + 1] = func
    end
    return {pressed = pressed, input = input, on = on}
end
return ____exports

File: src/game/core/input/Controller.ts
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

import { InputKeys } from "../../../ecs/CONSTS";


export function newController() {
    const listeners: { [key: number]: ((isPressed: boolean) => void)[] } = {} as { [key in typeof InputKeys[number]]: ((isPressed: boolean) => void)[] };
    const hashesKey = InputKeys.map(k => hash(k));
    function input(action_id: hash, action: any) {
        if (action_id != hash('touch') && (action.pressed || action.released))
            pressed(action_id, action.pressed != undefined && action.pressed == true);
    }

    function on(key: typeof InputKeys[number], func: (isPressed: boolean) => void) {
        if (listeners[InputKeys.indexOf(key)] == undefined)
            listeners[InputKeys.indexOf(key)] = [];
        listeners[InputKeys.indexOf(key)].push(func);
    }


    function pressed(key: hash, isPressed: boolean) {
        if (hashesKey.includes(key)) {
            if (listeners[hashesKey.indexOf(key)] != undefined) {
                for (const listener of listeners[hashesKey.indexOf(key)])
                    listener(isPressed);
            }
        }

    }

    return {
        pressed, input, on
    };
}